### PR TITLE
make deployment status behaviour more descriptive

### PR DESCRIPTION
**Pull request by atiratree**
There is sometimes a misunderstanding about the behaviour of Deployments and its  `Progressing` condition [1] [2]. This PR is trying to describe in more detail behaviour and peculiarities of the current implementation.

@ravisantoshgudimetla @soltysh what is your opinion on these changes? 

1. https://github.com/kubernetes/kubernetes/issues/106054 
2. https://github.com/kubernetes/kubernetes/issues/106697

TODO: 
- [ ] go over and update API definitions as well?

### CODE REVIEW

File: content/en/docs/concepts/workloads/controllers/deployment.md [hunk 1]
```diff
@@ -842,6 +842,13 @@ Kubernetes marks a Deployment as _progressing_ when one of the following tasks i
 * The Deployment is scaling down its older ReplicaSet(s).
 * New Pods become ready or available (ready for at least [MinReadySeconds](#min-ready-seconds)).
 
+When the rollout becomes progressing, the Deployment controller adds a DeploymentCondition with the following
+attributes to the Deployment's `.status.conditions`:
+
+* Type=Progressing
+* Status=True
+* Reason=NewReplicaSetCreated | FoundNewReplicaSet | ReplicaSetUpdated
```
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/deployment.md:850]: Although the existing text uses this style, I'd rather not see more of it. How about switching to something like:

```suggestion
* `type: Progressing`
* `status: "true"`
* `reason: NewReplicaSetCreated` | `reason: FoundNewReplicaSet` | `reason: ReplicaSetUpdated`
```

(I think) this is how they'd look in YAML, which is how we usually represent API details. JSON is also OK.

File: content/en/docs/concepts/workloads/controllers/deployment.md [hunk 2]
```diff
@@ -842,6 +842,13 @@ Kubernetes marks a Deployment as _progressing_ when one of the following tasks i
 * The Deployment is scaling down its older ReplicaSet(s).
 * New Pods become ready or available (ready for at least [MinReadySeconds](#min-ready-seconds)).
 
+When the rollout becomes progressing, the Deployment controller adds a DeploymentCondition with the following
```
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/deployment.md:845]:
```suggestion
When the rollout becomes “progressing”, the Deployment controller adds a condition with
the following
```
`DeploymentCondition` isn't part of the Kubernetes API; it's an implementation detail of the Golang implementation.
_In theory_ someone could drop in their own conformant Deployment controller and use that. We should write docs that are still relevant for that case.

File: content/en/docs/concepts/workloads/controllers/deployment.md [hunk 3]
```diff
@@ -853,6 +860,15 @@ updates you've requested have been completed.
 * All of the replicas associated with the Deployment are available.
 * No old replicas for the Deployment are running.
 
+When the rollout completes, the Deployment controller sets a DeploymentCondition with the following
+attributes to the Deployment's `.status.conditions`:
+
+* Type=Progressing
+* Status=True
+* Reason=NewReplicaSetAvailable
```
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/deployment.md:868]: see previous comment - let's try to show how things would look in the API.

File: content/en/docs/concepts/workloads/controllers/deployment.md [hunk 4]
```diff
@@ -853,6 +860,15 @@ updates you've requested have been completed.
 * All of the replicas associated with the Deployment are available.
 * No old replicas for the Deployment are running.
 
+When the rollout completes, the Deployment controller sets a DeploymentCondition with the following
```
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/deployment.md:863]: see previous comment - let's try to show how things would look in the API.

File: content/en/docs/concepts/workloads/controllers/deployment.md [hunk 5]
```diff
@@ -853,6 +860,15 @@ updates you've requested have been completed.
...
+attributes to the Deployment's `.status.conditions`:
+
+* Type=Progressing
+* Status=True
+* Reason=NewReplicaSetAvailable
+
+This condition will stay `True`, until a new rollout is initiated (even when availability of replicas changes, in that case see `Type=Available` condition).
```
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/deployment.md:870]:
```suggestion
This `Progressing` condition will retain a status value of `"True"` until a new rollout
is initiated. The condition holds even when availability of replicas changes (which
does instead affect the `Available` condition).
```
?

File: content/en/docs/concepts/workloads/controllers/deployment.md (hunk 1 above)
Comment (sftim) [reply content/en/docs/concepts/workloads/controllers/deployment.md:850]: Does the Deployment controller also emit an Event? If so, maybe we can mention that?

**Review by sftim — COMMENTED** [review]
Some feedback from a docs style perspective. It's not blocking. I do though want to push back somewhat on introducing more Go-isms into end user docs.

File: content/en/docs/concepts/workloads/controllers/deployment.md (hunk 1 above)
Comment (atiratree) [reply content/en/docs/concepts/workloads/controllers/deployment.md]: I would rather not in this case. 
1. It emits scaling event. Eg. `Scaled up replica set test to 1`. But this is not reflective of a rollout / deployment and occurs in plain scaling as well.
2. If you create Deployment with 0 replicas there is no event emitted.

**Comment by atiratree** [comment]
@sftim Thank you for the review. I have hopefully fixed all the comments.

I have switched the condition references to the YAML format and also applied this formatting to other places in this file.

**Review by soltysh — APPROVED** [review]
/lgtm

**Pull request merged**